### ISSUE TITLE
Plotting Surface Boundaries of Point Cloud objects resting on a Table

### ISSUE THREAD

**Issue by ishangupta3** [pyvista/pyvista-support#90]
I currently have 3 pointclouds (table, laptop, and grid) objects which I am merging into 1 single scene and resting the laptop and grid on the table. My goal is to extract the surface boundaries from the laptop and grid, and visualize these 2D surface boundaries directly on the table. I have tried extracting edges, and other things, but nothing really seems to work. 

I currently have this:

<img width="438" alt="Screen Shot 2019-12-19 at 1 39 21 PM" src="https://user-images.githubusercontent.com/8518915/71212048-5cab6880-2265-11ea-8165-f42aad94cc68.png">

2d Version of laptop pointcloud:
<img width="367" alt="Screen Shot 2019-12-19 at 1 48 13 PM" src="https://user-images.githubusercontent.com/8518915/71212486-481ba000-2266-11ea-815a-b835af42202d.png">

My goal is to have something like this:
![Screen Shot 2019-12-19 at 1 39 40 PM](https://user-images.githubusercontent.com/8518915/71212495-5073db00-2266-11ea-977a-cb44313b53ab.jpg)

My current Code:

    #Reading Table PointCloud
    surface = pv.read('extracted/table.ply')
    surface.points *= 30 # This makes the point cloud bigger

    #reading Laptop Point Cloud
    laptop = pv.read('extracted/laptop.ply')
    laptop.points *= 10
    laptop.translate([0, 3, 8])
    laptop.rotate_y(-90)

    #Getting Grid point cloud
    mesh_grid =  examples.load_uniform()
    mesh_grid.translate([-10, 0 ,-25]) #Translation example: https://github.com/pyvista/pyvista/blob/master/docs/core/points.rst
    mesh_grid.rotate_y(90)
    # edges_meshgrid = mesh_grid.extract_edges()
    # points = mesh_grid.points
    # cloud = pv.PolyData(points) # Converting numpy array into pyvista object
    # grid2d = cloud.delaunay_2d()
    # edges_grid2d = grid2d.extract_edges()

    #merging the 3 objects together
    merged = pv.PolyData(surface.points) # + pv.PolyData(mesh_grid.points) + pv.PolyData(laptop.points)
    merged.rotate_x(90)

    p = pv.Plotter()
    p.add_mesh(pv.PolyData(merged))
    p.show_grid()
    p.show()

**Comment by akaszynski** [comment]
Interesting problem.  The surfaces may not exactly intersect.  Would it be possible to upload the data here as a zip?

**Comment by ishangupta3** [comment]
I have attached a zip containing table.ply and laptop.ply, to place the laptop on the table, feel free to look at the code above. I also have other interesting .ply objects such as monitor, plant, that this problem can be applied to as well. 

[data2.zip](https://github.com/pyvista/pyvista-support/files/3985786/data2.zip)

**Comment by ishangupta3** [comment]
Yes, the surfaces probably do not intersect. For the most part I am interested in visualizing the surface boundaries on the table; it is alright if they do not exactly intersect. Any pointers on making them exactly intersect would be great as well.

**Comment by banesullivan** [comment]
This definitely is an interesting problem and dabbling a bit more into computer vision than computer graphics...i.e. something PyVista hasn't focused on much.

Interestingly, [this paper](https://ethz.ch/content/dam/ethz/special-interest/baug/igp/photogrammetry-remote-sensing-dam/documents/pdf/timo-jan-cvpr2016.pdf) implements a methodology for edge detection of point clouds that has been mostly implemented in Python in the [`pyntcloud`](https://github.com/daavoo/pyntcloud) library which has some connections with PyVista that need to be expanded.

Perhaps @daavoo might have some insight on how to do this?

Note: I'm a bit worried that the table point cloud might not be dense enough.

-----

reference: https://stackoverflow.com/questions/42600900/edge-detection-for-3d-point-clouds-meshes-with-r

**Comment by akaszynski** [comment]
I'm also having trouble getting an intersection edge to come up.  Normally, I'd try intersecting the two datasets, but generating a 3d delaunay doesn't give me a clean surface to work with as it generates many internal triangles.

This is one of those cases where I feel VTK tries really hard to do it all, but it simply can't.  It's primarily a visualization package, and while it certainly can do some mesh booleans, its boolean library functions can be quite brittle and it's not suited for all data sets.  While I can get close (see below), I think you're going to need a investigate another library (thanks @banesullivan) and see if its possible.

For you to do this properly in `pyvista`/`vtk`, you'll need manifold (or nearly manifold) data sets, which means you'll also need denser point clouds.

If you do figure out how to do this (either using `pyvista` or another library) please post back here.

```python
import pymeshfix

import pyvista as pv
from pyvista import examples

# Table point cloud to mesh
table = pv.read('table.ply')
table.points *= 30 # This makes the point cloud bigger
table = table.delaunay_3d(alpha=5.0).extract_surface()
# table = table.delaunay_3d(alpha=5.0)

# Laptop point cloud to mesh
laptop = pv.read('laptop.ply')
laptop.points *= 10
laptop.translate([0, 3, 8])
laptop.rotate_y(-90)
laptop = laptop.delaunay_3d(alpha=1.0).extract_surface()


#Getting Grid point cloud
mesh_grid =  examples.load_uniform()
mesh_grid.translate([-10, 0 ,-25]) #Translation example: https://github.com/pyvista/pyvista/blob/master/docs/core/points.rst
mesh_grid.rotate_y(90)

cpos = [(84.88003840730461, 50.76930344928887, -80.27498020925603),
        (0.023688793182373047, -0.9913382530212402, -0.191314697265625),
        (-0.34123374655088623, 0.9119452534969097, 0.22784947847055187)]

p = pv.Plotter()
p.add_mesh(laptop, color='w')
p.add_mesh(table, color='red')
p.camera_position = cpos
p.show_grid()
p.show()
```
![Selection_132](https://user-images.githubusercontent.com/11981631/71230178-ddf80f00-22e8-11ea-86d0-02f4e49d1a04.png)

**Comment by ishangupta3** [comment]
Thank you, what would be the process on PyVista to find the intersection of 2 datasets?
